src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -568,39 +902,31 @@ public void SetNonBlocking()
             }
         }
 
-        private bool TryBeginOperation<TOperation>(ref OperationQueue<TOperation> queue, TOperation operation, Interop.Sys.SocketEvents events, bool maintainOrder, out bool isStopped)
+        private void PerformSyncOperation<TOperation>(ref OperationQueue<TOperation> queue, TOperation operation, int timeout, int observedSequenceNumber)
             where TOperation : AsyncOperation
         {
-            // Exactly one of the two queue locks must be held by the caller
-            Debug.Assert(Monitor.IsEntered(_sendQueue.QueueLock) ^ Monitor.IsEntered(_receiveQueue.QueueLock));
-
-            switch (queue.State)
+            using (var e = new ManualResetEventSlim(false, 0))
             {
-                case QueueState.Stopped:
-                    isStopped = true;
-                    return false;
+                operation.Event = e;
 
-                case QueueState.Clear:
-                    break;
+                if (!queue.StartAsyncOperation(this, operation, observedSequenceNumber))
+                {
+                    // Completed synchronously
+                    return;
+                }
 
-                case QueueState.Set:
-                    if (queue.IsEmpty || !maintainOrder)
-                    {
-                        isStopped = false;
-                        queue.State = QueueState.Clear;
-                        return false;
-                    }
-                    break;
-            }
+                if (e.Wait(timeout))
+                {
+                    // Completed within timeout
+                    return;
+                }
 
-            if ((_registeredEvents & events) == Interop.Sys.SocketEvents.None)
-            {
-                Register(events);
+                bool cancelled = operation.TryCancel();
+                if (cancelled)
+                {
+                    operation.ErrorCode = SocketError.TimedOut;
+                }
```

Comment:
By this point, are we guaranteed that the operation will never try to Set the MRES?  I'm wondering because that's not obvious just from this code, and we dispose of the MRES on the next line.